internal/idp_test/idp_test.go

```
@@ -0,0 +1,89 @@
+package idp_test
+
+import (
+	"bytes"
+	"embed"
+	"encoding/json"
+	"fmt"
+	"io"
+	"io/fs"
+	"net/http/httptest"
+	"path/filepath"
+	"reflect"
+	"strings"
+	"testing"
+)
+
+//go:embed testdata
+var testdata embed.FS
+
+func TestIdP(t *testing.T) {
+	idPs, _ := testdata.ReadDir("testdata")
+	for _, idP := range idPs {
+		t.Run(idP.Name(), func(t *testing.T) {
+			idpPath := fmt.Sprintf("testdata/%s", idP.Name())
+			if err := fs.WalkDir(testdata, idpPath, func(path string, d fs.DirEntry, err error) error {
+				if err != nil {
+					return err
+				}
+				if d.IsDir() {
+					return nil
+				}
+				raw, err := fs.ReadFile(testdata, path)
+				if err != nil {
+					return fmt.Errorf("%s: %v", path, err)
+				}
+				var test testCase
+				if err := json.Unmarshal(raw, &test); err != nil {
+					return fmt.Errorf("%s: %v", path, err)
+				}
+				fileName, _ := filepath.Rel(idpPath, path)
+				t.Run(strings.TrimSuffix(fileName, ".json"), func(t *testing.T) {
+					if err := testRequest(test); err != nil {
+						t.Error(err)
+					}
+				})
+				return nil
```

You can just return the error of `testRequest` here, I think.